komposable-architecture-test/src/main/java/com/toggl/komposable/test/store/TestStore.kt

```diff
@@ -0,0 +1,482 @@
+package com.toggl.komposable.test.store
+
+import com.toggl.komposable.architecture.Effect
+import com.toggl.komposable.architecture.ReduceResult
+import com.toggl.komposable.architecture.Reducer
+import com.toggl.komposable.architecture.Store
+import com.toggl.komposable.extensions.createStore
+import com.toggl.komposable.extensions.map
+import com.toggl.komposable.scope.DispatcherProvider
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.BUFFERED
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.advanceTimeBy
+import kotlinx.coroutines.test.runCurrent
+import kotlinx.coroutines.yield
+import kotlin.random.Random
+import kotlin.reflect.full.memberProperties
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.TimeSource
+
+open class TestStore<State : Any, Action : Any?>(
+    initialState: () -> State,
+    reducer: () -> Reducer<State, Action>,
+    dispatcherProvider: DispatcherProvider,
+    private val logger: Logger,
+    private val reflectionHandler: ReflectionHandler = PublicPropertiesReflectionHandler(),
+    val testCoroutineScope: TestScope,
+    var timeout: Duration = 100.milliseconds,
+    var exhaustivity: Exhaustivity = Exhaustivity.Exhaustive,
+) {
+    private val store: Store<State, TestReducer.TestAction<Action>>
+    private val reducer: TestReducer<State, Action>
+    val state: State
+        get() = reducer.state
+
+    private val effectSubscriptionChannel = Channel<Unit>(capacity = BUFFERED)
+
+    init {
+        this.reducer = TestReducer(
+            reducer(),
+            initialState(),
+            testCoroutineScope,
+            effectSubscriptionChannel,
+        )
+        this.store = createStore(
+            initialState = this.state,
+            reducer = this.reducer,
+            dispatcherProvider = dispatcherProvider,
+            storeScopeProvider = { testCoroutineScope },
+        )
+    }
+
+    private val assertionRunner: AssertionRunner<State, Action>
+        get() = exhaustivity.let {
+            when (it) {
+                is Exhaustivity.Exhaustive -> ExhaustiveAssertionRunner()
+                is Exhaustivity.NonExhaustive -> NonExhaustiveAssertionRunner(
+                    it.logIgnoredStateChanges,
+                    it.logIgnoredReceivedActions,
+                    it.logIgnoredEffects,
+                )
+            }
+        }
```

Comment:
Perhaps more readable? And should work the same? 
```suggestion
        get() = when (val e = exhaustivity) {
                is Exhaustivity.Exhaustive -> ExhaustiveAssertionRunner()
                is Exhaustivity.NonExhaustive -> NonExhaustiveAssertionRunner(
                    e.logIgnoredStateChanges,
                    e.logIgnoredReceivedActions,
                    e.logIgnoredEffects,
                )
            }
        }
```

Reply:
Actually, how about we extract this mapping to some extension method on Exhaustivity? 

`Exhaustivity.toAssertionRunner()` or something. The idea is to untangle TestStore from the AssertionRunner implementations a bit more. We could even put them into a separate file along with the extension function.

Reply:
Moved it to the new file as suggested.